feat(#11): 질문 상세조회 페이지 UI 개발

jngmnj opened the pull request:
## 🔧 관련 이슈

- 이 PR은 다음 이슈와 관련 있습니다: #11

## 🔄 변경 사항

- [x] 기능 추가

## ✔️ 변경 사항 상세 설명

- **MarkdownRenderer 컴포넌트 (`src/components/common/MarkdownEditor/MarkdownRenderer.tsx`)**
  - 마크다운 텍스트를 HTML로 렌더링하는 컴포넌트 개발.
  - `remark-gfm` 및 `rehype-raw` 플러그인 적용.
  - **코드 블록 및 인라인 코드 렌더링 로직 개선:**
    - `components.pre`를 통해 모든 **블록 코드** (언어 명시/미명시 무관)를 `react-syntax-highlighter`로 처리.
    - `components.code`는 `<pre>` 태그로 감싸지지 않은 **순수 인라인 코드**만을 담당하도록 역할 분리.
    - `inline` prop의 불확실성을 우회하고 `<pre>` 태그 존재 여부로 코드 유형을 정확히 구분하여 스타일 적용.
    - `SyntaxHighlighter`에 `vscDarkPlus` 스타일 적용.
    - 불필요한 `highlight.js` 스타일 임포트 제거.

- **Avatar 컴포넌트 (`src/components/common/Avatar.tsx`)**
  - 사용자 프로필 이미지를 표시하는 아바타 컴포넌트 구현.
  - `name` (alt 텍스트) 및 `profileUrl` props를 통해 이미지 렌더링.
  - 이미지 로드 실패 시 `handleImageError` 유틸리티를 사용하여 기본 아바타 이미지로 대체.

- **formatRelativeTime 유틸리티 (`src/utils/formatRelativeTime.ts`)**
  - 특정 시점으로부터 현재까지의 시간을 '방금 전', 'X분 전', 'X시간 전' 등 상대적인 형태로 포맷팅하는 함수 추가.

- **getFullDate 유틸리티 (`src/utils/getFullDate.ts`)**
  - ISO 날짜 문자열을 'YYYY년 MM월 DD일' 형식으로 포맷팅하는 함수 추가.

- **handleImageError 유틸리티 (`src/utils/handleImageError.ts`)**
  - 이미지 로드 에러 발생 시 `type`에 따라 일반 기본 이미지 또는 기본 아바타 이미지로 대체하는 공통 유틸리티 함수 구현.

- **AnswerCard 컴포넌트 (`src/components/qna/AnswerCard.tsx`)**
  - 단일 답변 내용을 표시하는 카드 컴포넌트 개발.
  - `MarkdownRenderer`를 통해 답변 본문 마크다운 렌더링.
  - 답변 작성자 정보, 질문자 채택 여부, 댓글 목록 및 입력 폼 포함.
  - 댓글 정렬 (`최신순/오래된 순`) 기능 구현.
  - `Avatar`, `formatRelativeTime`, `getFullDate` 유틸리티 활용.

- **QnaDetailPage 컴포넌트 (`src/pages/QnaDetailPage.tsx`)**
  - Q&A 상세 페이지 레이아웃 및 기능 구현.
  - 질문 내용, 카테고리, 작성자 정보, 조회수, 작성일 표시.
  - `AnswerCard` 컴포넌트를 사용하여 답변 목록 렌더링.
  - 페이지 URL 공유하기 (`navigator.clipboard`) 기능 및 토스트 메시지 연동.

## 📸 스크린샷 (UI 변경 시 필수)
![localhost_5173_qna_1_](https://github.com/user-attachments/assets/3acdd8d5-6059-4661-9c02-a05d8ef2a02f)

## 📝 문서화

- [ ] 관련 문서가 업데이트되었습니다. (`README.md`, Notion 링크 등)

jngmnj committed:
💡 chore: MarkdownRenderer 컴포넌트 패키지 추가(#67)

- highlight.js
- react-markdown
- rehype-highlight
- rehype-raw

jngmnj committed:
✨ feat: MarkdownRenderer 컴포넌트 개발(#67)

- `react-markdown`을 사용하여 마크다운 콘텐츠를 렌더링하는 컴포넌트 추가
- `remark-gfm` 플러그인으로 GitHub Flavored Markdown (GFM) 지원
- `rehype-raw` 플러그인으로 HTML 삽입 지원
- 기본 코드 블록 및 인라인 코드 스타일 정의

jngmnj committed:
✨ feat: 상대적 시간 포맷팅 유틸리티 추가(#11)

- `formatRelativeTime` 함수를 개발하여 특정 시점으로부터의 상대적 시간 표시
- '방금 전', 'X분 전', 'X시간 전', 'X일 전', 'X달 전', 'X년 전' 형식 지원

jngmnj committed:
✨ feat: 이미지 로드 에러 처리 유틸리티 추가(#11)

- `handleImageError` 함수를 개발하여 이미지 로드 실패 시 대체 이미지 표시
- 아바타 이미지와 일반 이미지에 대한 기본 대체 이미지 설정 지원

jngmnj committed:
✨ feat: 날짜 전체 포맷팅 유틸리티 추가(#11)

- `getFullDate` 함수를 개발하여 ISO 문자열을 'YYYY년 MM월 DD일' 형식으로 변환
- 사용자 친화적인 날짜 표시를 위한 포맷팅 기능 제공

jngmnj committed:
✨ feat: Avatar 컴포넌트 개발(#11)

- 사용자 프로필 이미지를 표시하는 `Avatar` 컴포넌트 구현
- `name`과 `profileUrl` props를 받아 이미지 렌더링
- `handleImageError` 유틸리티를 사용하여 이미지 로드 실패 시 기본 아바타 이미지로 대체
- Tailwind CSS를 활용한 원형 이미지 및 오버플로우 처리

jngmnj committed:
✨ feat: Q&A 상세 페이지 및 답변 카드 컴포넌트 개발(#11)

- `QnaDetailPage` 컴포넌트를 구현하여 질문 상세 내용, 답변 목록 표시
- `AnswerCard` 컴포넌트를 구현하여 개별 답변 내용 및 댓글 표시
- 질문자 채택, 상대적 시간(`formatRelativeTime`), 전체 날짜(`getFullDate`) 표시 기능 통합
- `Avatar`, `MarkdownRenderer`, `Textarea` 등 재사용 가능한 컴포넌트 활용
- 댓글 입력 폼 및 댓글 목록 렌더링 로직 포함
- Q&A 공유하기 기능 (`navigator.clipboard`) 추가 및 토스트 메시지 연동

jngmnj committed:
♻️  refactor: 테스트용 마크다운 파일 임포트 제거 및 더미 데이터 정리(#11)

- `QnaDetailPage` 컴포넌트 내에서 테스트 목적으로 사용되던 `answer_ex1.md` 및 `answer_ex2.md` 파일 임포트 문 삭제.
- 실제 프로젝트에 포함되지 않는 테스트용 `.md` 파일 참조로 인한 잠재적 빌드 오류 방지.
- 필요시 컴포넌트 내부에서 직접 더미 마크다운 문자열을 사용하여 테스트 진행.

Aprasaks reviewed (dismissed):
pull 땅겨오면 npm i 해야겠네요 
수고많으셨습니다!

devjaesung commented on src/components/common/MarkdownEditor/MarkdownRenderer.tsx:
```diff
@@ -0,0 +1,50 @@
+  className?: string
+}
+
+const MarkdownRenderer: React.FC<MarkdownRendererProps> = ({
+  content,
```
React.FC 타입은 주의해서 사용해야 한다는 점도 알고 계시면 좋을 것 같습니다!
[참고1](https://shape-coding.tistory.com/entry/TypeScript-ReactFC%EC%97%90-%EC%82%AC%EC%9A%A9%EC%97%90-%EB%8C%80%ED%95%B4-%EC%83%9D%EA%B0%81%ED%95%B4%EB%B3%B4%EA%B8%B0)
[참고2](https://www.inflearn.com/community/questions/1113860/%ED%83%80%EC%9E%85%EC%8A%A4%ED%81%AC%EB%A6%BD%ED%8A%B8%EB%A1%9C-%EB%A6%AC%EC%95%A1%ED%8A%B8%EB%A5%BC-%ED%94%84%EB%A1%9C%EC%A0%9D%ED%8A%B8%EB%A5%BC-%ED%95%98%EB%A9%B4%EC%84%9C-react-fc%EB%9D%BC%EB%8A%94-%EA%B2%83%EC%97%90-%EC%A7%88%EB%AC%B8%EC%9D%B4-%EC%9E%88%EC%8A%B5%EB%8B%88%EB%8B%A4?srsltid=AfmBOooSagz9Az1mBPTH2UJfqE-QjFJr-Xfp32xKscd7aZTDanoUU9_B)

devjaesung commented on src/components/qna/AnswerCard.tsx:
```diff
@@ -0,0 +1,126 @@
+          </button>
+        </form>
+        {/* 댓글 목록 */}
+        {comments.map((comment, idx) => (
+          <div key={idx} className="flex items-start gap-4 mt-5">
```
현재 map을 comments를 대상으로 적용하고 있어서 최신순 정렬이 적용되지 않을거 같아요..! 기능 추가를 나중에 하실 생각이라면 참고해주세요!

devjaesung replied:
댓글이 없을 때의 분기 처리도 추가되면 좋을 것 같네요..!

devjaesung reviewed (dismissed):
정민님 수고하셨습니다!! 파일 분리가 잘되서 너무 깔끔하네요..!

jngmnj committed:
✨ feat: 질문 상세 페이지 라우트 추가(#11)

jngmnj committed:
♻️  refactor: AnswerCard 컴포넌트 기능 추가 및 개선(#11)

- 댓글 최신/오래된 순 정렬기능
- 댓글 정렬버튼 hover style 추가
- className 삭제(오타+불필요)
- 댓글 없을경우 분기처리
- 댓글 item key 수정(comment.id)

jngmnj committed:
♻️  refactor: MarkdownRenderer 컴포넌트 수정(#67)

- React.FC<T> 사용지양으로, interface에 children직접추가하여 타이핑

jngmnj commented:
피드백하신부분 수정했습니다!

- MarkdownRenderer 컴포넌트 수정
  - React.FC<T> 사용지양으로, interface에 children직접추가하여 타이핑
- AnswerCard 컴포넌트 기능 추가 및 개선
  - 댓글 최신/오래된 순 정렬기능
  - 댓글 정렬버튼 hover style 추가
  - className 삭제(오타+불필요)
  - 댓글 없을경우 분기처리
  - 댓글 item key 수정(comment.id)

Aprasaks reviewed (approved):
수고많으셨어요. 파일깔끔해영!

chen4023 commented on src/types/qnaDetail.ts:
```diff
@@ -0,0 +1,23 @@
+  content: string
+  is_adopted: boolean
+  created_at: string
+  author: {
+    id: number
+    nickname: string
+    profile_image_url: string
+  }
```
Author 타입을 개별로 지정해서 사용해도 좋을 것 같네요!

chen4023 commented on src/pages/QnaDetailPage.tsx:
```diff
@@ -0,0 +1,191 @@
+        ],
+      },
+    ],
+  }
```
목업 데이터는 폴더로 분리해서 관리해주시면 좋을 것 같아요!

이후 실제 API를 연결하는 과정에서 파일 별로 지우는 작업을 방지하기 위해서입니다!

chen4023 commented on src/components/qna/AnswerCard.tsx:
```diff
@@ -0,0 +1,136 @@
+    e.preventDefault()
+  }
+
+  const sortedComments = [...comments].sort((a, b) => {
+    const aTime = new Date(a.created_at).getTime()
+    const bTime = new Date(b.created_at).getTime()
+    return orderByDesc ? bTime - aTime : aTime - bTime
+  })
```
컴포넌트가 렌더링될 때 마다 정렬연산이 수행되기 때문에 `useMemo`를 사용하여 성능적인 측면을 고려해보시는 것도 좋을 것 같네요!

chen4023 commented on src/components/common/MarkdownEditor/MarkdownRenderer.tsx:
```diff
@@ -0,0 +1,51 @@
+      ...props
+    }: ComponentProps<'code'> & { inline?: boolean }) {
+      return inline ? (
+        <code className="bg--100 px-1 rounded text-sm">{children}</code>
```
확인 부탁드려요!

jngmnj replied:
수정하겠습니다!!!

chen4023 reviewed (approved):
전체적으로 잘 구현해주셨네요 정민님!
참고하실 코멘트 몇가지 남겨두어, 참고해주시면 좋을 것 같아요 :)
고생 많으셨습니다!

yunsung-hodoopapa commented on src/components/common/Avatar.tsx:
```diff
@@ -0,0 +1,23 @@
+        src={profileUrl}
+        alt={name}
+        className="w-full h-full object-cover"
+        onError={(e) => handleImageError(e, 'avatar')}
```
Goood

yunsung-hodoopapa commented on src/components/qna/AnswerCard.tsx:
```diff
@@ -0,0 +1,136 @@
+      </div>
+    </div>
+  )
+}
```
이 안에서만 사용되는 유틸함수는 
```ts
function something() { 
...
};
```

yunsung-hodoopapa commented on src/components/qna/AnswerCard.tsx:
```diff
@@ -0,0 +1,136 @@
+  const { author, created_at, content, comments, is_adopted } = answer
+  const borderColor = is_adopted ? 'border-primary' : 'border-gray-250'
+
+  const toggleOrderBy = () => {
```
이벤트 핸들러에 넘기는 함수
onClick = > handle -

yunsung-hodoopapa commented on src/pages/QnaDetailPage.tsx:
```diff
@@ -0,0 +1,191 @@
+      toast.show({ message: 'URL이 복사되었습니다!', type: 'success' })
+    } catch {
+      toast.show({ message: 'URL 복사 실패', type: 'error' })
+    }
```
catch {
...
} finally {
 //마지막에 실행시키고 싶은 로직
}

yunsung-hodoopapa commented on src/pages/QnaDetailPage.tsx:
```diff
@@ -0,0 +1,191 @@
+    }
+  }
+
+  const qnaData = {
```
_mocks_ <-

yunsung-hodoopapa commented on src/utils/formatRelativeTime.ts:
```diff
@@ -0,0 +1,19 @@
+export function formatRelativeTime(createdAt: string | Date): string {
```
dateFns 혹은 day.js 패키지를 사용하시면 아주아주 쉽게 구현하실 수 있습니다.

yunsung-hodoopapa replied:
주의사항 // moment.js
